### PR TITLE
Update performance-ptbr.md

### PR DESCRIPTION
**Pull request by dialmformart**
Adding a new part of the translation.

@danielfireman :D

### CODE REVIEW

File: performance-ptbr.md [hunk 1]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
 
 Para transferência de dados, vá para um protocolo menos falador ou aumente a API para permitir consultas parciais. Por exemplo, uma consulta incremental em vez de ser
 forçado a buscar o conjunto de dados inteiro a cada vez.
+
+### Alterações algorítmicas
+
+Se você não estiver alterando os dados, a outra opção principal é alterar o código.
+
+A maior melhoria provavelmente virá de uma alteração algorítmica. Isso equivale a substituir um bubble sort (`O(n^2)`)  por um quicksort (`O(n log n)`) ou substituir uma varredura linear de uma matriz (`O(n)`) por uma pesquisa binária (`O (log n)`) ou uma pesquisa de mapa (`O(1)`).
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
A maior melhoria provavelmente virá de uma alteração algorítmica. Isso equivale a substituir um bubble sort (`O(n^2)`)  por um quicksort (`O(n log n)`) ou substituir a varredura linear de uma matriz (`O(n)`) por uma pesquisa binária (`O (log n)`) ou pesquisa em um mapa (`O(1)`).
```

File: performance-ptbr.md [hunk 2]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+É assim que o software se torna lento. As estruturas projetadas originalmente para um uso são reaproveitadas para algo para o qual não foram projetadas. Isso acontece gradualmente.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
É assim que o software se torna lento. As estruturas projetadas originalmente para um uso são reaproveitadas para algo que não foram projetadas. Isso acontece gradualmente.
```

File: performance-ptbr.md [hunk 3]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+
+É importante ter uma compreensão intuitiva dos diferentes níveis de grandeza O.
+Escolha a estrutura de dados certa para o seu problema. Você não precisa sempre cortar os ciclos, mas isso apenas evita problemas de desempenho burros que podem não ser percebidos até muito mais tarde.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Escolha a estrutura de dados certa para o seu problema. Você não precisa cortar ciclos sempre, mas isso evita problemas de desempenho bobos que podem não ser percebidos até muito mais tarde.
```

File: performance-ptbr.md [hunk 4]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+### Alterações algorítmicas
+
+Se você não estiver alterando os dados, a outra opção principal é alterar o código.
+
+A maior melhoria provavelmente virá de uma alteração algorítmica. Isso equivale a substituir um bubble sort (`O(n^2)`)  por um quicksort (`O(n log n)`) ou substituir uma varredura linear de uma matriz (`O(n)`) por uma pesquisa binária (`O (log n)`) ou uma pesquisa de mapa (`O(1)`).
+
+É assim que o software se torna lento. As estruturas projetadas originalmente para um uso são reaproveitadas para algo para o qual não foram projetadas. Isso acontece gradualmente.
+
+
+É importante ter uma compreensão intuitiva dos diferentes níveis de grandeza O.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
É importante ter uma compreensão intuitiva dos diferentes níveis Grande-O.
```

Mais abaixo, você chama a notação de Grande-O

File: performance-ptbr.md [hunk 5]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Escolha a estrutura de dados razoável mais simples e siga em frente. Este é o CS 101 para escrever "software não lento". 
+Esse deve ser o seu modo de desenvolvimento padrão. Se você sabe que precisa de acesso aleatório, não escolha uma lista vinculada.
+Se você sabe que precisa de uma travessia em ordem, não use um mapa. 
+Os requisitos mudam e você nem sempre pode adivinhar o futuro. Faça um palpite razoável sobre a carga de trabalho.
+
+<http://daslab.seas.harvard.edu/rum-conjecture/>
+
+As estruturas de dados para problemas semelhantes diferem quando executam um trabalho. Uma árvore binária é classificada um pouco
+de cada vez à medida que as inserções acontecem. Uma matriz não ordenada é mais rápida de inserir, mas não é ordenada: no final para "finalizar", você precisa fazer a ordenação de uma só vez.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
de cada vez à medida que as inserções acontecem. Uma matriz não ordenada é mais rápida de inserir, mas não é ordenada: ao final, para "finalizar", você precisa fazer a ordenação de uma só vez.
```

File: performance-ptbr.md [hunk 6]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Ao escrever um pacote para ser usado por outras pessoas, evite a tentação de otimizar antecipadamente todos os casos de uso. Isso resultará em código ilegível. As estruturas de dados por design são efetivamente de propósito único. Você não pode ler mentes nem prever o futuro. Se um usuário disser "Seu pacote está muito lento para este caso de uso", uma resposta razoável pode ser "Então use este outro pacote aqui". Um pacote deve "fazer uma coisa bem".
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Ao escrever um pacote para ser usado por outras pessoas, evite a tentação de otimizar antecipadamente todos os casos de uso. Isso resultará em código ilegível. Por projeto, as estruturas de dados são efetivamente de propósito único. Você não pode ler mentes nem prever o futuro. Se um usuário disser "Seu pacote está muito lento para este caso de uso", uma resposta razoável pode ser "Então use este outro pacote aqui". Um pacote deve "fazer uma coisa bem".
```

File: performance-ptbr.md [hunk 7]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Às vezes, estruturas de dados híbridas fornecem a melhoria de desempenho que você precisa. Por exemplo, ao reunir seus dados, você pode limitar sua pesquisa a um único intervalo. Isso ainda paga o custo teórico de O(n), mas a constante será menor. Revisitaremos esses tipos de ajustes quando chegarmos ao ajuste do programa.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Às vezes, estruturas de dados híbridas fornecem a melhoria de desempenho que você precisa. Por exemplo, ao reunir seus dados, você pode limitar sua pesquisa a um único intervalo. Isso ainda paga o custo teórico de O(n), mas a constante será menor. Revisitaremos esses tipos de ajustes quando chegarmos em otimizações.
```

File: performance-ptbr.md [hunk 8]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Duas coisas que as pessoas esquecem quando discutem a notação grande-O:
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Duas coisas que as pessoas esquecem quando discutem a notação Grande-O:
```

Ou colocar maiúsculo aqui ou padronizar minúsculo no resto do texto

File: performance-ptbr.md [hunk 9]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Um, há um fator constante envolvido. Dois algoritmos que têm a mesma complexidade algorítmica podem ter diferentes fatores constantes. Imagine repetir uma lista 100 vezes ou apenas repetir uma vez. Embora ambos sejam O (n), um tem um fator constante 100 vezes maior.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Primeiramente, há um fator constante envolvido. Dois algoritmos que têm a mesma complexidade algorítmica podem ter diferentes fatores constantes. Imagine repetir uma lista 100 vezes ou apenas repetir uma vez. Embora ambos sejam O (n), um tem um fator constante 100 vezes maior.
```

File: performance-ptbr.md [hunk 10]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Esses fatores constantes são o motivo pelo qual, embora a classificação por mesclagem, a classificação rápida e a classificação geral de todos os O (nlogn), todos usem a classificação rápida porque é a mais rápida. Tem o menor fator constante.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Esses fatores constantes são o motivo pelo qual, embora merge sort, quick sort e heap sort sejam O (nlogn), todo mundo usa quick sort porque é a mais rápida. Este método de ordenação tem o menor fator constante.
```

File: performance-ptbr.md [hunk 11]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+A segunda coisa é que Grande-O diz apenas "à medida que n cresce até o infinito". Ele fala sobre a tendência de crescimento: "À medida que os números aumentam, esse é o fator de crescimento que dominará o tempo de execução". Não diz nada sobre o desempenho real, ou como ele se comporta com um valor de n pequeno.
+
+Freqüentemente há um ponto de corte abaixo do qual um algoritmo mais burro é mais rápido. Um bom exemplo do pacote sort da biblioteca padrão Go. Na maioria das vezes, ele usa o quicksort, mas ele passa por uma classificação por shell e depois por inserção quando o tamanho da partição cai abaixo de 12 elementos.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Freqüentemente há um ponto de corte abaixo do qual um algoritmo  ingênuo é mais rápido. Um bom exemplo do pacote sort da biblioteca padrão Go. Na maioria das vezes, ele usa o quicksort, mas ele passa por uma classificação por shell e depois por inserção quando o tamanho da partição cai abaixo de 12 elementos.
```

File: performance-ptbr.md [hunk 12]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Para alguns algoritmos, o fator constante pode ser tão grande que esse ponto de corte pode ser maior que todas as entradas razoáveis. Ou seja, o algoritmo O(n^2) é mais rápido que o algoritmo O(n) para todas as entradas com as quais você provavelmente lida.
+
+
+Isso também significa que você precisa conhecer os tamanhos de entrada representativos, tanto para escolher o algoritmo mais apropriado quanto para escrever boas avaliações. 10 itens? 1000 itens? 1000000 itens?
+
+Isso também acontece de outra maneira: por exemplo, optar por usar uma estrutura de dados mais complicada para fornecer o escalonamento de O (n) em vez de O (n^2),
+mesmo que com os parâmetros de referência para pequenas entradas tenham ficado mais lentos. Isso também se aplica à maioria das estruturas de dados sem bloqueio.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
mesmo que com os parâmetros de referência para pequenas entradas tenham ficado mais lentos. Isso também se aplica à maioria das estruturas de dados sem contenção. 
```

File: performance-ptbr.md [hunk 13]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Para alguns algoritmos, o fator constante pode ser tão grande que esse ponto de corte pode ser maior que todas as entradas razoáveis. Ou seja, o algoritmo O(n^2) é mais rápido que o algoritmo O(n) para todas as entradas com as quais você provavelmente lida.
+
+
+Isso também significa que você precisa conhecer os tamanhos de entrada representativos, tanto para escolher o algoritmo mais apropriado quanto para escrever boas avaliações. 10 itens? 1000 itens? 1000000 itens?
+
+Isso também acontece de outra maneira: por exemplo, optar por usar uma estrutura de dados mais complicada para fornecer o escalonamento de O (n) em vez de O (n^2),
+mesmo que com os parâmetros de referência para pequenas entradas tenham ficado mais lentos. Isso também se aplica à maioria das estruturas de dados sem bloqueio. 
+Eles geralmente são mais lentos no caso de uma única thread, mas são mais escaláveis quando muitas threads o estão usando.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Elas geralmente são mais lentas no caso de uma única thread, mas são mais escaláveis quando estão sendo usadas por muitas threads.
```

File: performance-ptbr.md [hunk 14]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+A hierarquia de memória nos computadores modernos confunde um pouco o problema aqui, pois os caches preferem o acesso previsível da varredura de uma fatia ao acesso efetivamente aleatório da perseguição de um ponteiro. Ainda assim, é melhor começar com um bom algoritmo. Falaremos sobre isso na seção específica de hardware.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
A hierarquia de memória nos computadores modernos confunde um pouco o problema aqui, isto pois os caches preferem o acesso previsível de uma varredura ao acesso aleatório que temos com a perseguição de um ponteiro. Ainda assim, é melhor começar com um bom algoritmo. Falaremos sobre isso na seção específica de hardware.
```

File: performance-ptbr.md [hunk 15]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+TODO: estendendo o último parágrafo, mencione a notação O () é um modelo em que cada 
+operação tem custo fixo. Essa é uma suposição errada no hardware moderno.
+
+> A luta pode nem sempre ser a mais forte, nem a corrida a mais rápida,
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
> A luta nem sempre é vencida pelo mais forte, nem a corrida pelo mais rápido,
```

File: performance-ptbr.md [hunk 16]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+mas essa é a maneira de apostar.
+> -- <cite>Rudyard Kipling</cite>
+
+Às vezes, o melhor algoritmo para um problema específico não é um único algoritmo, mas uma coleção de algoritmos especializados para classes de entrada ligeiramente diferentes. Esse "polialgoritmo" detecta rapidamente com que tipo de entrada ele precisa lidar e despacha para o caminho de código apropriado. É isso que o pacote de classificação mencionado acima: determina o tamanho do problema e escolhe um algoritmo diferente. Além de combinar quicksort, classificação de shell,
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Às vezes, o melhor algoritmo para um problema específico não é um único algoritmo, mas uma coleção de algoritmos especializados para classes de entrada ligeiramente diferentes. Esse "polialgoritmo" detecta rapidamente com que tipo de entrada ele precisa lidar e despacha para o caminho de código apropriado. É isso que faz o pacote de classificação mencionado acima: determina o tamanho do problema e escolhe um algoritmo diferente. Além de combinar quicksort, classificação de shell,
```

File: performance-ptbr.md [hunk 17]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+e classificação de inserção, também rastreia a profundidade de recursão do quicksort e chama heapsort, se necessário. Os pacotes `string` e` bytes` fazem algo semelhante, detectando e se especializando para casos diferentes. Assim como na compactação de dados, quanto mais você souber sobre a aparência de sua entrada, melhor poderá ser sua solução personalizada. Mesmo que uma otimização nem sempre seja aplicável, vale a pena complicar seu código, determinando que é seguro usar e executar lógicas diferentes.
+
+Isso também se aplica aos subproblemas que seu algoritmo precisa resolver. 
+Por exemplo, poder usar a classificação radix pode ter um impacto significativo no desempenho ou
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Por exemplo, ter o usar radix sort a disposição pode ter um impacto significativo no desempenho ou 
```

File: performance-ptbr.md [hunk 18]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Algoritmos mais simples também podem ser mais rápidos. Esses dois exemplos não são casos isolados
+  https://go-review.googlesource.com/c/crypto/+/169037
+  https://go-review.googlesource.com/c/go/+/170322/
+
+TODO: notas sobre seleção de algoritmo
+
+TODO:
+  melhore o comportamento do pior caso a um custo leve para o tempo de execução médio
+  correspondência de Expressão regular de tempo linear
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
  da correspondência de expressão regular de tempo linear
```

File: performance-ptbr.md [hunk 19]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+
+Embora a maioria dos algoritmos seja determinísticp, há uma classe de algoritmos que usam a aleatoriedade como uma maneira de simplificar etapas de tomada de decisão complexas.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Embora a maioria dos algoritmos seja determinística, há uma classe de algoritmos que usam a aleatoriedade como uma maneira de simplificar etapas de tomada de decisão complexas.
```

File: performance-ptbr.md [hunk 20]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+Em vez de ter um código que faz a coisa certa, você usa a aleatoriedade para selecionar uma coisa provavelmente não *ruim*. Por exemplo, um treap é uma árvore binária probabilisticamente equilibrada. Cada nó tem uma chave, mas também recebe um valor aleatório. 
+Ao inserir na árvore, o caminho de inserção normal da árvore binária é seguido, mas os nós também obedecem à propriedade heap
+com base no peso atribuído aleatoriamente a cada nó. Essa abordagem mais simples substitui soluções de 
+rotação de árvores complicadas (como árvores AVL e Rubro negras), mas ainda mantém uma árvore equilibrada com inserção/pesquisa de O (log n) "com alta
+probabilidade". As Skip lists são outra estrutura de dados simples e semelhante que usa aleatoriedade para produzir "provavelmente" inserção e pesquisas de O (log n).
+
+Da mesma forma, a escolha de um pivô aleatório para quicksort pode ser mais simples do que uma abordagem de média mediana mais complexa para encontrar um bom pivô, 
+e a probabilidade de maus pivôs serem continuamente escolhidos (aleatoriamente) e degradar o desempenho do quicksort para O(n^2) é muito pequeno.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
e a probabilidade de maus pivôs serem continuamente escolhidos (aleatoriamente) e degradar o desempenho do quicksort para O(n^2) é muito pequena.
```

File: performance-ptbr.md [hunk 21]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Os algoritmos aleatórios são classificados como algoritmos "Monte Carlo" ou "Las Vegas", depois de dois locais de jogo bem conhecidos.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Os algoritmos aleatórios são classificados como algoritmos "Monte Carlo" ou "Las Vegas", a partir de dois locais de jogo bem conhecidos. 
```

File: performance-ptbr.md [hunk 22]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+Como vimos, algoritmos diferentes fazem sentido em diferentes tamanhos de entrada.
+Se o seu intervalo de entrada esperado for <100, seus benchmarks devem refletir isso. Caso contrário, escolher um algoritmo ideal para n = 10^6 pode não ser o mais rápido.
+
+Ser capaz de gerar dados de teste representativos. Diferentes distribuições de dados podem provocar comportamentos diferentes em seu algoritmo: 
+pense no exemplo clássico de "quicksort é O (n^2) quando os dados são classificados". Da mesma forma, a pesquisa de interpolação é O (log log n) para dados aleatórios 
+uniformes, mas O (n) no pior caso. Saber como são as suas entradas é a chave para os benchmarks representativos e para escolher o melhor algoritmo. 
+Se os dados que você está usando para testar não são representativos de cargas de trabalho reais, você pode facilmente otimizar um determinado conjunto de dados,
+"ajustando demais" seu código para funcionar melhor com um conjunto específico de entradas.
+
+Isso também significa que seus dados de referência precisam ser representativos dos reais
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Isso também significa que seus dados de referência precisam ser representativos do mundo real.
```

File: performance-ptbr.md [hunk 23]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+mundo. O uso de entradas puramente aleatórias pode distorcer o comportamento do seu algoritmo.
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
 O uso de entradas puramente aleatórias pode distorcer o comportamento do seu algoritmo.
```

File: performance-ptbr.md [hunk 24]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+Os algoritmos de cache e compactação exploram distribuições distorcidas ausentes
+em dados aleatórios e, portanto, terá um desempenho pior, enquanto uma árvore binária executará
+melhor com valores aleatórios, pois eles tendem a manter a árvore equilibrada. (Isto é
+a ideia por trás de uma armadilha, a propósito.)
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
a ideia por trás de uma treap, a propósito.)
```

File: performance-ptbr.md [hunk 25]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+
+Por outro lado, considere o caso de testar um sistema com um cache. Se seu
+A entrada de benchmark consiste apenas em uma única consulta, então cada solicitação atingirá o
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Se sua entrada benchmark consiste apenas em uma única consulta, então cada solicitação atingirá o
```

File: performance-ptbr.md [hunk 26]
```diff
@@ -257,3 +257,211 @@ Outro aspecto a considerar é o tempo de transferência de dados. Geralmente, o
...
+"ajustando demais" seu código para funcionar melhor com um conjunto específico de entradas.
+
+Isso também significa que seus dados de referência precisam ser representativos dos reais
+mundo. O uso de entradas puramente aleatórias pode distorcer o comportamento do seu algoritmo.
+Os algoritmos de cache e compactação exploram distribuições distorcidas ausentes
+em dados aleatórios e, portanto, terá um desempenho pior, enquanto uma árvore binária executará
+melhor com valores aleatórios, pois eles tendem a manter a árvore equilibrada. (Isto é
+a ideia por trás de uma armadilha, a propósito.)
+
+Por outro lado, considere o caso de testar um sistema com um cache. Se seu
```
Comment (danielfireman) [inline performance-ptbr.md]:
```suggestion
Por outro lado, considere o caso de testar um sistema com um cache. 
```

**Review by dialmformart — COMMENTED** [review]
Fixed!

**Comment by dialmformart** [comment]
@dgryski Can you merge?

@danielfireman already reviewed it and I fixed it with the improvements he said

**Comment by dgryski** [comment]
Thanks!

**Pull request merged**